AnnService/inc/Core/Common/cuda/KNN.hxx

```diff
@@ -366,591 +251,220 @@ __global__ void findRNG_strict(Point<T,SUMTYPE,Dim>* data, TPtree<T,KEY_T,SUMTYP
         }
       }
       for(size_t j=0; j<KVAL; j++) {
-        results[(size_t)(query.id-min_id)*KVAL+j] = threadList[j].idx;
+        results[(size_t)(queryId-min_id)*KVAL+j] = threadList[j].idx;
       }
-
     } // End if within batch
   } // End leaf node loop
 }
 
-/*****************************************************************************************
- * Perform the brute-force graph construction on each leaf node, but tries to maintain 
- * RNG properties when adding/removing vectors from the neighbor list.  Also only
- * inserts new vectors into the neighbor list when the "accessibility score" is 0 (i.e.,
- * it is not already accessible by any neighbors.
- *****************************************************************************************/
-template<typename T, typename KEY_T, typename SUMTYPE, int Dim, int BLOCK_DIM>
-__global__ void findRNG_leaf_nodes(Point<T,SUMTYPE,Dim>* data, TPtree<T,KEY_T,SUMTYPE,Dim>* tptree, int KVAL, int* results, int metric) {
 
-  extern __shared__ char sharememory[];
-  ThreadHeap<T, SUMTYPE, Dim, BLOCK_DIM> heapMem; // Stored in registers but contains a pointer 
+#define RUN_KERNEL(size)          \
+  if(dim <= size) {               \
+    DistPair<SUMTYPE>* threadList = (&((DistPair<SUMTYPE>*)sharememory)[KVAL*threadIdx.x]); \
+    SUMTYPE (*dist_comp)(T*,T*);    \
+    dist_comp = &dist<T,SUMTYPE,size,metric>;  \
+    findRNG<T,SUMTYPE,size>(ps, tptree, KVAL, results, min_id, max_id, threadList, dist_comp); \
+    return; \
+  } 
 
-  // Assigns the memory pointers in the heap to shared memory
-  heapMem.initialize(&((DistPair<SUMTYPE>*)sharememory)[(KVAL-1) * threadIdx.x], KVAL-1);
+#define RUN_KERNEL_QUANTIZED(size) \
+  if(dim <= size) {\
+    DistPair<float>* threadList = (&((DistPair<float>*)sharememory)[KVAL*threadIdx.x]); \
+    findRNG_PQ<size>((PointSet<uint8_t>*)ps, tptree, KVAL, results, min_id, max_id, threadList, quantizer); \
+    return; \
+  } 
 
-  SUMTYPE nearest_dist;
-  Point<T,SUMTYPE,Dim> query;
+#define MAX_SHAPE 1024
```

Comment:
To clarify, this is the max dimension supported, versus old approach where each supported dimension is added individually?

Reply:
And RUN_KERNEL and RUN_TAIL_KERNEL compile the specific non-PQ template dimensions, and RUN_KERNEL_QUANTIZED and RUN_TAIL_KERNEL_PQ do the same for PQ?

Reply:
Exactly. The MAX_SHAPE is the largest currently supported, but that could be changed easily if higher dimensions are needed.  Adding the specific dimensions of the dataset to the RUN_KERNEL will improve performance, but if it is not, the next highest dimension will be used.